File: WeeklyCode/week5/Path Sum/gaoqiang.py
# Definition for a  binary tree node
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    # @param root, a tree node
    # @param sum, an integer
    # @return a boolean
    targetValue = 0
    def hasPathSum(self, root, sum):
    	self.targetValue = sum
    	return self.dfs(root,0)

    def dfs(self, root, sum):
    	if root==None:
    		return False
    	sum = sum+root.val
    	if sum==self.targetValue and root.left==None and root.right==None:
    		return True
    	result = self.dfs(root.left, sum)
    	if result:
    		return result
    	result = self.dfs(root.right, sum)
    	return result
        

File: WeeklyCode/week11/Combinations/gaoqiang.java
public class Solution {
    private List<List<Integer>> result = null;
    public List<List<Integer>> combine(int n, int k) {
        result = new ArrayList<List<Integer>>();
        if(n<=0||k<=0||n<k){
            return result;
        }
        combineNext(n, k, 1, new ArrayList<Integer>());
        return result;
    }
    
    public void combineNext(int n, int k, int current, List<Integer> currentList){
        if(currentList.size()==k){
            result.add(currentList);
            return;
        }
        if(current>n||(currentList.size() +(n-current+1))<k){
            return;
        }
        List<Integer> left = new ArrayList<Integer>(), right = new ArrayList<Integer>();
        left.addAll(currentList);
        right.addAll(currentList);
        right.add(current);
        combineNext(n,k,current+1,left);
        combineNext(n,k,current+1,right);
    }
}